File: src/main/java/de/bixilon/minosoft/protocol/packets/clientbound/play/PacketMultiBlockChange.java
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.protocol.packets.clientbound.play;

import de.bixilon.minosoft.data.mappings.blocks.Block;
import de.bixilon.minosoft.data.world.ChunkLocation;
import de.bixilon.minosoft.data.world.InChunkLocation;
import de.bixilon.minosoft.logging.Log;
import de.bixilon.minosoft.protocol.packets.ClientboundPacket;
import de.bixilon.minosoft.protocol.protocol.InByteBuffer;
import de.bixilon.minosoft.protocol.protocol.PacketHandler;

import java.util.HashMap;

public class PacketMultiBlockChange implements ClientboundPacket {
    final HashMap<InChunkLocation, Block> blocks = new HashMap<>();
    ChunkLocation location;

    @Override
    public boolean read(InByteBuffer buffer) {
        if (buffer.getVersionId() < 25) {
            if (buffer.getVersionId() < 4) {
                this.location = new ChunkLocation(buffer.readVarInt(), buffer.readVarInt());
            } else {
                this.location = new ChunkLocation(buffer.readInt(), buffer.readInt());
            }
            short count = buffer.readShort();
            int dataSize = buffer.readInt(); // should be count * 4
            for (int i = 0; i < count; i++) {
                int raw = buffer.readInt();
                byte meta = (byte) (raw & 0xF);
                short blockId = (short) ((raw & 0xFF_F0) >>> 4);
                byte y = (byte) ((raw & 0xFF_00_00) >>> 16);
                byte z = (byte) ((raw & 0x0F_00_00_00) >>> 24);
                byte x = (byte) ((raw & 0xF0_00_00_00) >>> 28);
                this.blocks.put(new InChunkLocation(x, y, z), buffer.getConnection().getMapping().getBlockById((blockId << 4) | meta));
            }
            return true;
        }
        if (buffer.getVersionId() < 740) {
            this.location = new ChunkLocation(buffer.readInt(), buffer.readInt());
            int count = buffer.readVarInt();
            for (int i = 0; i < count; i++) {
                byte pos = buffer.readByte();
                byte y = buffer.readByte();
                int blockId = buffer.readVarInt();
                this.blocks.put(new InChunkLocation((pos & 0xF0 >>> 4) & 0xF, y, pos & 0xF), buffer.getConnection().getMapping().getBlockById(blockId));
            }
            return true;
        }
        long rawPos = buffer.readLong();
        this.location = new ChunkLocation((int) (rawPos >> 42), (int) (rawPos << 22 >> 42));
        int yOffset = ((int) rawPos & 0xFFFFF) * 16;
        if (buffer.getVersionId() > 748) {
            buffer.readBoolean(); // ToDo
        }
        int count = buffer.readVarInt();
        for (int i = 0; i < count; i++) {
            long data = buffer.readVarLong();
            this.blocks.put(new InChunkLocation((int) ((data >> 8) & 0xF), yOffset + (int) ((data >> 4) & 0xF), (int) (data & 0xF)), buffer.getConnection().getMapping().getBlockById(((int) (data >>> 12))));
        }
        return true;
    }

    @Override
    public void handle(PacketHandler h) {
        h.handle(this);
    }

    @Override
    public void log() {
        Log.protocol(String.format("[IN] Multi block change received at %s (size=%d)", this.location, this.blocks.size()));
    }

    public ChunkLocation getLocation() {
        return this.location;
    }

    public HashMap<InChunkLocation, Block> getBlocks() {
        return this.blocks;
    }
}

File: src/main/java/de/bixilon/minosoft/protocol/packets/clientbound/play/PacketMultiBlockChange.kt
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */
package de.bixilon.minosoft.protocol.packets.clientbound.play

import de.bixilon.minosoft.data.mappings.blocks.BlockState
import de.bixilon.minosoft.data.mappings.tweaker.VersionTweaker
import de.bixilon.minosoft.gui.rendering.util.VecUtil.inChunkSectionPosition
import de.bixilon.minosoft.gui.rendering.util.VecUtil.sectionHeight


import de.bixilon.minosoft.modding.event.events.MultiBlockChangeEvent
import de.bixilon.minosoft.protocol.network.Connection
import de.bixilon.minosoft.protocol.packets.ClientboundPacket
import de.bixilon.minosoft.protocol.protocol.InByteBuffer
import de.bixilon.minosoft.protocol.protocol.ProtocolVersions
import de.bixilon.minosoft.util.logging.Log
import glm_.vec2.Vec2i
import glm_.vec3.Vec3i
import java.util.*

class PacketMultiBlockChange : ClientboundPacket() {
    val blocks = HashMap<Vec3i, BlockState?>()
    lateinit var chunkPosition: Vec2i
        private set

    override fun read(buffer: InByteBuffer): Boolean {
        if (buffer.versionId < ProtocolVersions.V_14W26C) {
            chunkPosition = if (buffer.versionId < ProtocolVersions.V_1_7_5) {
                Vec2i(buffer.readVarInt(), buffer.readVarInt())
            } else {
                buffer.readChunkPosition()
            }
            val count = buffer.readShort()
            val dataSize = buffer.readInt() // should be count * 4
            for (i in 0 until count) {
                val raw = buffer.readInt()
                val meta = (raw and 0xF)
                val blockId = (raw and 0xFFF0 ushr 4)
                val y = (raw and 0xFF0000 ushr 16)
                val z = (raw and 0x0F000000 ushr 24)
                val x = (raw and -0x10000000 ushr 28)
                blocks[Vec3i(x, y, z)] = buffer.connection.mapping.getBlockState((blockId shl 4) or meta)
            }
            return true
        }
        if (buffer.versionId < ProtocolVersions.V_20W28A) {
            chunkPosition = Vec2i(buffer.readInt(), buffer.readInt())
            val count = buffer.readVarInt()
            for (i in 0 until count) {
                val position = buffer.readByte().toInt()
                val y = buffer.readByte()
                val blockId = buffer.readVarInt()
                blocks[Vec3i(position and 0xF0 ushr 4 and 0xF, y.toInt(), position and 0xF)] = buffer.connection.mapping.getBlockState(blockId)
            }
            return true
        }
        val rawPos = buffer.readLong()
        chunkPosition = Vec2i((rawPos shr 42).toInt(), (rawPos shl 22 shr 42).toInt())
        val yOffset = (rawPos.toInt() and 0xFFFFF) * 16
        if (buffer.versionId > ProtocolVersions.V_1_16_2_PRE3) {
            buffer.readBoolean() // ToDo
        }
        val count = buffer.readVarInt()
        for (i in 0 until count) {
            val data = buffer.readVarLong()
            blocks[Vec3i((data shr 8 and 0xF).toInt(), yOffset + (data shr 4 and 0xF).toInt(), (data and 0xF).toInt())] = buffer.connection.mapping.getBlockState((data ushr 12).toInt())
        }
        return true
    }

    override fun handle(connection: Connection) {
        val chunk = connection.player.world.getChunk(chunkPosition) ?: return // thanks mojang
        if (!chunk.isFullyLoaded) {
            return
        }
        connection.fireEvent(MultiBlockChangeEvent(connection, this))
        chunk.setRawBlocks(blocks)

        // tweak
        if (!connection.version.isFlattened()) {
            for ((key, value) in blocks) {
                val block = VersionTweaker.transformBlock(value!!, chunk.sections!!, key.inChunkSectionPosition, key.sectionHeight)
                if (block === value) {
                    continue
                }
                chunk.setBlockState(key, block)
            }
        }
        val sectionHeights = HashSet<Int>()
        for ((key) in blocks) {
            sectionHeights.add(key.sectionHeight)
        }
        for (sectionHeight in sectionHeights) {
            connection.renderer.renderWindow.worldRenderer.prepareChunkSection(chunkPosition, sectionHeight)
        }
    }

    override fun log() {
        Log.protocol(String.format("[IN] Multi block change received at %s (size=%d)", chunkPosition, blocks.size))
    }
}
